src/memray/_memray/tracking_api.cpp

```diff
@@ -248,18 +251,53 @@ std::atomic<bool> Tracker::d_active = false;
 std::unique_ptr<Tracker> Tracker::d_instance_owner;
 std::atomic<Tracker*> Tracker::d_instance = nullptr;
 MEMRAY_FAST_TLS thread_local size_t NativeTrace::MAX_SIZE{64};
+thread_local BernoulliSampler Tracker::d_sampler{};
```

Comment:
Hm. There's only two piece of mutable state inside the `BernoulliSampler` - the `random_engine`, and `d_bytes_until_next_sample`. It looks like we can easily make the `random_engine` thread-local, since (at least on glibc) it has a trivial destructor, so we don't need to worry about using it while a thread is being torn down. If we make the `random_engine` thread-local, I think we can use a `std::atomic` for `d_bytes_until_next_sample`, to allow sharing a single sampler across all threads without needing any mutexes.

This implementation seems to pass the tests, after making `d_sampler` a regular instance variable instead of a `static thread_local` - can you poke any holes in this?

```cpp
class BernoulliSampler
{
  public:
    // Methods
    void setSamplingInterval(size_t sampling_interval);
    size_t calculateSampleSize(size_t size);

  private:
    // Methods
    size_t poissonStep();

    // Data members
    size_t d_sampling_interval_in_bytes;
    double d_sampling_probability;
    std::atomic<size_t> d_bytes_until_next_sample;
};

void
BernoulliSampler::setSamplingInterval(size_t sampling_interval)
{
    d_sampling_interval_in_bytes = sampling_interval;
    d_sampling_probability = 1.0 / d_sampling_interval_in_bytes;
    d_bytes_until_next_sample = poissonStep();
}

size_t
BernoulliSampler::poissonStep()
{
    static_assert(std::is_trivially_destructible<std::default_random_engine>::value);
    static thread_local std::default_random_engine s_random_engine;
    std::exponential_distribution<double> exponential_dist(d_sampling_probability);
    return static_cast<size_t>(exponential_dist(s_random_engine)) + 1;
}

size_t
BernoulliSampler::calculateSampleSize(size_t size)
{
    if (size >= d_sampling_interval_in_bytes) {
        return size;
    }

    size_t n_samples = 0;
    size_t next_step = 0;
    while (size) {
        size_t loaded_bytes_until_next_sample = d_bytes_until_next_sample.load();
        if (size >= loaded_bytes_until_next_sample) {
            // We're being sampled! We need to replace our atomic entirely.
            // Randomly decide when to stop next, if we haven't already.
            if (!next_step) {
                next_step = poissonStep();
            }

            if (!d_bytes_until_next_sample.compare_exchange_weak(
                        loaded_bytes_until_next_sample,
                        next_step)) {
                continue;  // Didn't update the atomic; loop and try again.
            }

            // Successfully updated the atomic; some bytes are accounted for.
            next_step = 0;  // Our random number has been used
            size -= loaded_bytes_until_next_sample;
            n_samples++;
        } else {
            // We're not being sampled; just decrement our atomic.
            if (!d_bytes_until_next_sample.compare_exchange_weak(
                        loaded_bytes_until_next_sample,
                        loaded_bytes_until_next_sample - size))
            {
                continue;  // Didn't update the atomic; loop and try again.
            }

            // Successfully updated the atomic; all bytes are accounted for.
            size = 0;
        }
    }

    return d_sampling_interval_in_bytes * n_samples;
}
```

Reply:
I think it would be possible to use `memory_order_acq_rel` for the read/modify/write, and `memory_order_relaxed` for the loads - but I don't claim to understand the C++ memory model terribly well, and that may be premature optimization.

Reply:
In fact, I _think_ it would be correct to use `memory_order_relaxed` for everything, since we're only looking for atomicity and don't care at all about any writes any other thread may have done to any address besides our atomic.